language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
     }
     spec remove_identities {
         let account_addr = Signer::spec_address_of(account);
-        let num_identities = len(global<NetworkIdentity>(account_addr).identities);
-        // aborts_if len(to_remove) == 0;
-        // aborts_if len(to_remove) > MAX_ADDR_IDENTITIES;
+        let prior_identities = global<NetworkIdentity>(account_addr).identities;
+        let has_change = (exists i in 0..len(to_remove): contains(prior_identities, to_remove[i]));
+
+        let post handle = global<NetworkIdentity>(account_addr).identity_change_events;
+        let post msg = NetworkIdentityChangeNotification {
+            identities: global<NetworkIdentity>(account_addr).identities,
+            time_rotated_seconds: DiemTimestamp::spec_now_seconds(),
+        };
+
+        aborts_if len(to_remove) == 0;
+        aborts_if len(to_remove) > MAX_ADDR_IDENTITIES;
+        aborts_if !exists<NetworkIdentity>(account_addr);
+        include has_change ==> DiemTimestamp::AbortsIfNotOperating;
+        include RemoveMembersInternalEnsures<vector<u8>> {
+            old_members: prior_identities,
+            new_members: global<NetworkIdentity>(account_addr).identities,
+        };
         modifies global<NetworkIdentity>(account_addr);
-        invariant exists<NetworkIdentity>(account_addr);
-        invariant num_identities <= MAX_ADDR_IDENTITIES;
+        emits msg to handle if has_change;
     }
 
-    /// Removes an identity and returns true if a change was made
-    fun remove_identity(identities: &mut vector<vector<u8>>, to_remove: vector<u8>): bool {
-        let (exist, i) = Vector::index_of(identities, &to_remove);
+    // =================================================================
+    // Set operation simulation
+
+    /// Add all elements that appear in `to_add` into `members`.
+    ///
+    /// The `members` argument is essentially a set simulated by a vector, hence
+    /// the uniqueness of its elements are guaranteed, before and after the bulk
+    /// insertion. The `to_add` argument, on the other hand, does not guarantee
+    /// to be a set and hence can have duplicated elements.
+    fun add_members_internal<T: copy>(
```

Comment:
We should probably just make a `Set` type somewhere in the language, since now it's separated out :)

Reply:
Yes, we are actually discussing this (we have a thread on the move-prover channel).